docs/proposals/image-index.md

```diff
@@ -0,0 +1,150 @@
+# Image Classification API Proposal
+
+## Problem
+
+Allow to efficiently report usage of images in the cluster by tracking the
+relationship between images from the **base** image to the **application** image.
+
+By having this system in place, we can provide a better auditing of which and
+how images are used. In case of a vulnerability disclosure in one of the base
+images, we can identify the affected applications built on top of this base
+image and trigger an update.
+
+Additionally, administrator should be able to limit the usage to certain image
+classes (iow. On this cluster, you are only allowed to deploy applications
+based on *base* image).
+
+## Use Cases
+
+* Classify a certain image by a label and have that label propagated to all descentants of that
+  image currently present in the cluster.
+  * By classification, we are giving an Image context, or role in the cluster. For example an image
+    classified as "VENDOR" represents all images that are based on this "VENDOR" image.
+* Be able to query:
+  * ImageStreamTags that currently reference an Image classified as *foo*
+  * ImageStreamTags that reference outdated version of an Image classified as *foo*
+  * All Images based on the Image *foo* (descentants)
+  * All Images that the *foo* images is based on (ancestors)
+* Be able to track the usage of images limited by an license by giving the image carrying the licenced
+  content a class
+* Be able to enforce policies based on classification:
+  * Allow running only *VENDOR* based images in this cluster
+* Be able to perform bulk operations based on the classification:
+  * Trigger a build for all BuildConfigs that have builder image based on *VENDOR*.
```

Comment:
Some other ones now that relate to image availability:

1. From a master, efficiently determine whether a given pod should have access to an image by its digest, by mapping digest -> all image streams in the current namespace

Reply:
2. From the registry, perform image layer and blob checks on a namespace in O(1) to check permission
3. Allow a mirroring controller to find all possible image streams that may be able to import a given layer (layer -> image -> image stream) so that the controller can backfill images that need to be mirrored

Reply:
> From the registry, perform image layer and blob checks on a namespace in O(1) to check permission

@smarterclayton No. We can use this index to check because Oleg is rights and this mechanism produces false positives.

> Allow a mirroring controller to find all possible image streams that may be able to import a given layer (layer -> image -> image stream)

@smarterclayton This index does not allow you to track such relationships. This index is not about layers at all (layer is not unique).

This index allows you to establish with some probability what images contain other images. This is based on the sequence of layers, but they are not directly used. We move from them to the chains. It is important to understand: сhains hash for some layer addresses all the previous layers as well.

Reply:
Ok, then my use cases (while part of image indexing in general) are more specifically scoped to questions around fast layer overlap / access / use.  I think this case indexing may end up looking  more like a solitary controller, while a layer index may be something exposed via the main image api.  

From a use case perspective, O(1) ACL checks (simplify registry), O(1) presence checks (which images import this layer), and the ability for a controller to scan a list of layers quickly based on some properties of images (like say an annotation that says that this image was completely mirrored) in order to do mirroring.

We know that the registry is going to have challenges holding all image metadata in memory, so I'm looking for exploration of constructs we'll need to support use cases here clients have difficulty holding everything in memory.